cal/cal.1
.\"
.\" Sccsid @(#)cal.1	1.4 (gritter) 1/11/03
.\"
.\" Derived from cal(1), Unix 32V:
.\"
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.TH CAL 1 "1/11/03" "Heirloom Toolchest" "User Commands"
.SH NAME
cal \- print calendar
.SH SYNOPSIS
\fBcal\fR [[\fImonth\fR] \fIyear\fR]
.SH DESCRIPTION
.I Cal
prints a calendar for the specified year.
If a month is also specified, a calendar
just for that month is printed.
If no year is given,
the calendar for the current month is printed.
.I Year
can be between 1
and 9999.
The
.I month
is a number between 1 and 12.
The calendar
produced is that for England and her colonies.
.PP
Try September 1752.
.SH "ENVIRONMENT VARIABLES"
.TP
\fBLANG\fR, \fBLC_ALL\fR, \fBLC_CTYPE\fR, \fBLC_TIME\fR
Determines the month and weekday strings.
.SH "SEE ALSO"
calendar(1),
date(1),
locale(7)
.SH NOTES
The year is always considered to start in January even though this
is historically naive.
.br
Beware that `cal 78' refers to the early Christian era,
not the 20th century.


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

cal/cal.c
/*	from Unix 32V /usr/src/cmd/cal.c	*/
/*
 * Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
 *
 * SPDX-Licence-Identifier: Caldera 
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)cal.sl	1.12 (gritter) 5/29/05";

#include	<unistd.h>
#include	<time.h>
#include	<stdio.h>
#include	<libgen.h>
#include	<stdlib.h>
#include	<locale.h>
#include	<langinfo.h>
#include	<string.h>
#include	<wchar.h>
#include	<limits.h>

static wchar_t	dayw[] = {
	' ', 'S', ' ', ' ', 'M', ' ', 'T', 'u', ' ', ' ', 
	'W', ' ', 'T', 'h', ' ', ' ', 'F', ' ', ' ', 'S', 0
};
static char	*smon[]= {
	"January", "February", "March", "April",
	"May", "June", "July", "August",
	"September", "October", "November", "December"
};
static char	*amon[] = {
	"Jan", "Feb", "Mar", "Apr",
	"May", "Jun", "Jul", "Aug",
	"Sep", "Oct", "Nov", "Dec"
};
static char	string[432];
static char	*progname;

static void	usage(void);
static void	bad(void);
static void	dolocale(void);
static int	number(const char *str);
static void	pstr(char *str, int n);
static void	cal(int m, int y, char *p, int w);
static int	jan1(int yr);
static void	wput(const wchar_t *);

int
main(int argc, char **argv)
{
	register int y = 1, i, j;
	int m = 1;

	progname = basename(argv[0]);
	dolocale();
	if (argc > 1 && argv[1][0] == '-' && argv[1][1] == '-'
			&& argv[1][2] == '\0')
		argv++, argc--;
	if (argc == 2)
		goto xlong;
	else if (argc == 1) {
		time_t t;
		struct tm *tp;

		time(&t);
		tp = localtime(&t);
		m = tp->tm_mon + 1;
		y = tp->tm_year + 1900;
	} else if (argc == 3) {
		/*
 		 *	print out just month
 		 */
		m = number(argv[optind]);
		if(m<1 || m>12)
			bad();
		y = number(argv[optind + 1]);
		if(y<1 || y>9999)
			bad();
	} else
		usage();
	printf("   %s %u\n", smon[m-1], y);
	wput(dayw);
	putchar('\n');
	cal(m, y, string, 24);
	for(i=0; i<6*24; i+=24)
		pstr(string+i, 24);
	exit(0);

/*
 *	print out complete year
 */

xlong:
	y = number(argv[optind]);
	if(y<1 || y>9999)
		bad();
	printf("\n\n\n");
	printf("				%u\n", y);
	printf("\n");
	for(i=0; i<12; i+=3) {
		for(j=0; j<6*72; j++)
			string[j] = '\0';
		printf("	 %.3s", amon[i]);
		printf("			%.3s", amon[i+1]);
		printf("		       %.3s\n", amon[i+2]);
		wput(dayw);
		printf("   ");
		wput(dayw);
		printf("   ");
		wput(dayw);
		putchar('\n');
		cal(i+1, y, string, 72);
		cal(i+2, y, string+23, 72);
		cal(i+3, y, string+46, 72);
		for(j=0; j<6*72; j+=72)
			pstr(string+j, 72);
	}
	printf("\n\n\n");
	return 0;
}

static int
number(const char *str)
{
	register int n, c;
	register const char *s;

	n = 0;
	s = str;
	while((c = *s++) != '\0') {
		if(c<'0' || c>'9')
			return(0);
		n = n*10 + c-'0';
	}
	return(n);
}

static void
pstr(char *str, int n)
{
	register int i;
	register char *s;

	s = str;
	i = n;
	while(i--)
		if(*s++ == '\0')
			s[-1] = ' ';
	i = n+1;
	while(i--)
		if(*--s != ' ')
			break;
	s[1] = '\0';
	printf("%s\n", str);
}

char	mon[] = {
	0,
	31, 29, 31, 30,
	31, 30, 31, 31,
	30, 31, 30, 31,
};

static void
cal(int m, int y, char *p, int w)
{
	register int d, i;
	register char *s;

	s = p;
	d = jan1(y);
	mon[2] = 29;
	mon[9] = 30;

	switch((jan1(y+1)+7-d)%7) {

	/*
	 *	non-leap year
	 */
	case 1:
		mon[2] = 28;
		break;

	/*
	 *	1752
	 */
	default:
		mon[9] = 19;
		break;

	/*
	 *	leap year
	 */
	case 2:
		;
	}
	for(i=1; i<m; i++)
		d += mon[i];
	d %= 7;
	s += 3*d;
	for(i=1; i<=mon[m]; i++) {
		if(i==3 && mon[m]==19) {
			i += 11;
			mon[m] += 11;
		}
		if(i > 9)
			*s = i/10+'0';
		s++;
		*s++ = i%10+'0';
		s++;
		if(++d == 7) {
			d = 0;
			s = p+w;
			p = s;
		}
	}
}

/*
 *	return day of the week
 *	of jan 1 of given year
 */

static int
jan1(int yr)
{
	register int y, d;

/*
 *	normal gregorian calendar
 *	one extra day per four years
 */

	y = yr;
	d = 4+y+(y+3)/4;

/*
 *	julian calendar
 *	regular gregorian
 *	less three days per 400
 */

	if(y > 1800) {
		d -= (y-1701)/100;
		d += (y-1601)/400;
	}

/*
 *	great calendar changeover instant
 */

	if(y > 1752)
		d += 3;

	return(d%7);
}

static void
usage(void)
{
	fprintf(stderr, "usage: %s [ [month] year ]\n", progname);
	exit(2);
}

static void
bad(void)
{
	fprintf(stderr, "Bad argument\n");
	exit(1);
}

#define	COPY_DAY(d)		cp = nl_langinfo(DAY_##d); \
				cp += mbtowc(wq++, cp, MB_CUR_MAX); \
				cp += mbtowc(wq++, cp, MB_CUR_MAX); \
				wq++

#define	COPY_MONTH(m)		smon[m - 1] = strdup(nl_langinfo(MON_##m))

#define	COPY_ABMON(m)		amon[m - 1] = strdup(nl_langinfo(ABMON_##m))

static void
dolocale(void)
{
	char *cp;
	wchar_t *wq;

#ifndef	__dietlibc__
	setlocale(LC_CTYPE, "");
	cp = setlocale(LC_TIME, "");
	if (cp && strcmp(cp, "C") && strcmp(cp, "POSIX") && strcmp(cp, "en") &&
			strncmp(cp, "en_", 3)) {
#else	/* __dietlibc__ */
	if (1) {
#endif	/* __dietlibc__ */
		wq = dayw;
		COPY_DAY(1);
		COPY_DAY(2);
		COPY_DAY(3);
		COPY_DAY(4);
		COPY_DAY(5);
		COPY_DAY(6);
		COPY_DAY(7);
		COPY_MONTH(1);
		COPY_MONTH(2);
		COPY_MONTH(3);
		COPY_MONTH(4);
		COPY_MONTH(5);
		COPY_MONTH(6);
		COPY_MONTH(7);
		COPY_MONTH(8);
		COPY_MONTH(9);
		COPY_MONTH(10);
		COPY_MONTH(11);
		COPY_MONTH(12);
		COPY_ABMON(1);
		COPY_ABMON(2);
		COPY_ABMON(3);
		COPY_ABMON(4);
		COPY_ABMON(5);
		COPY_ABMON(6);
		COPY_ABMON(7);
		COPY_ABMON(8);
		COPY_ABMON(9);
		COPY_ABMON(10);
		COPY_ABMON(11);
		COPY_ABMON(12);
	}
}

static void
wput(const wchar_t *s)
{
	char	mb[MB_LEN_MAX];
	int	i, n;

	s--;
	while (*++s) {
		n = wctomb(mb, *s);
		for (i = 0; i < n; i++)
			putchar(mb[i] & 0377);
	}
}
